### ISSUE TITLE
ptr-ptr cast does not work well with lifetimes (lifetime widening)

### ISSUE THREAD

**Issue by arielb1** [rust-lang/rust#27214]
## STR

``` Rust
fn foo<'a>(v: *const &'a ()) {
    let v: *const &'static () = v as *const &'static ();
}

fn main() {}
```
## Expected result

cast will be successful
## Actual Result

```
<anon>:2:33: 2:34 error: mismatched types:
 expected `*const &'static ()`,
    found `*const &'a ()`
(lifetime mismatch) [E0308]
<anon>:2     let v: *const &'static () = v as *const &'static ();
                                         ^
<anon>:2:33: 2:34 help: see the detailed explanation for E0308
<anon>:1:30: 3:2 note: the lifetime 'a as defined on the block at 1:29...
<anon>:1 fn foo<'a>(v: *const &'a ()) {
<anon>:2     let v: *const &'static () = v as *const &'static ();
<anon>:3 }
note: ...does not necessarily outlive the static lifetime
error: aborting due to previous error
playpen: application terminated with error code 101
```

**Comment by steveklabnik** [comment]
Why would this cast work? `'a` can be shorter than `'static`.

**Comment by bluss** [comment]
Raw pointer casts are normally not restricted. `*const A as *const B` is usally ok, without any checking.

**Comment by steveklabnik** [comment]
Ah, I see, thanks @bluss . Seems reasonable.

**Comment by ranma42** [comment]
Does Rust guarantee that pointers are correctly aligned? Would it be legal in a safe context to create a misaligned pointer?
Neither https://doc.rust-lang.org/book/raw-pointers.html nor https://doc.rust-lang.org/book/unsafe.html mention any constraint on alignment.

**Comment by bluss** [comment]
Currently, loads and writes through raw pointer dereferences assume the pointer is aligned (this is guarded by `unsafe`).

`Box::new(t)` creates a `0x1` pointer if t is zero size, even if it needs more than 1-alignment. For example if it is a `[u64; 0]` (size 0, align of 8). From the box you can thus get a misaligned reference, however it is one that can't be read/written through (can't read/write something of 0 size).

For guarantees, have to ask https://internals.rust-lang.org/c/language-design/unsafe-code-guidelines

**Comment by Rufflewind** [comment]
As @talchas on `#rust` pointed out to me, `as` seems to forbid lifetime-only changes, but if anything else changes it's totally on board:

~~~rust
fn foo<'a>(v: *const &'a ()) {
    let v: *const &'static () = v as *const () as *const &'static ();
}
~~~

**Comment by steveklabnik** [comment]
Triage: still reproduces.

**Comment by fmease** [comment]
Triage: No longer reproduces (fail→pass). We could just close this as completed but I'm actually not sure which PR(s) was responsible. Maybe NLL? Might try to bisect later.

**Comment by fmease** [comment]
Bisects to 1.75 / nightly-2023-10-27 / PR #113262.
See also issue #113257 which duplicates this one.

Closing as fixed by #113262. No additional tests needed.
